server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java

```diff
@@ -0,0 +1,205 @@
+    }
+
+    @Test
+    public void testNullArrayNullObject() throws Exception {
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("The array argument of the `array_append` function can be of " +
+                                        "undefined type or of undefined inner type, but not when object argument " +
+                                        "is also undefined");
+        assertNormalize("array_append(null, null)", null);
+    }
+
+    @Test
+    public void testNullArrayInnerTypeNullObject() throws Exception {
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("The array argument of the `array_append` function can be of " +
+                                        "undefined type or of undefined inner type, but not when object argument " +
+                                        "is also undefined");
+        assertNormalize("array_append([null], null)", null);
+    }
+
+    @Test
+    public void testZeroArguments() throws Exception {
+        expectedException.expect(UnsupportedOperationException.class);
+        expectedException.expectMessage("Unknown function: array_append()." +
+                                        " Possible candidates: array_append(array(E), E):array(E)");
+        assertEvaluate("array_append()", null);
+    }
+
+    @Test
+    public void testOneArgument() {
+        expectedException.expect(UnsupportedOperationException.class);
+        expectedException.expectMessage("Unknown function: array_append(_array(1))," +
+                                        " no overload found for matching argument types: (integer_array).");
+        assertEvaluate("array_append([1])", null);
+    }
+
+    @Test
+    public void testThreeArguments() throws Exception {
+        expectedException.expect(UnsupportedOperationException.class);
+        expectedException.expectMessage("Unknown function: array_append(_array(1), _array(2), _array(3))," +
+                                        " no overload found for matching argument types: (integer_array, integer_array, integer_array).");
+        assertEvaluate("array_append([1], [2], [3])", null);
+    }
+
+    @Test
+    public void testNullVarArray() throws Exception {
+        assertEvaluate("array_append(int_array, 1)",
+                       List.of(1),
+                       Literal.of(INTEGER_ARRAY_TYPE, null));
+    }
+
+    @Test
+    public void testVarArrayWithNullObject() throws Exception {
+        assertEvaluate("array_append(int_array, null)",
+            Arrays.asList(1, null, 3, null),
+            Literal.of(Arrays.asList(1, null, 3), INTEGER_ARRAY_TYPE));
+    }
+
+    @Test
+    public void testTwoIntegerArguments() throws Exception {
+        assertEvaluate("array_append(int_array, 3)",
+            Arrays.asList(1, 2, 3),
+            Literal.of(Arrays.asList(1, 2), INTEGER_ARRAY_TYPE));
+    }
```

Comment:
Based on existing scalar test cases it may be a bit confusing when and if at all to test with `asserNormalize` vs  `assertEvaluate`. Actually this is only needed if the scalar implements a concrete `normalizeSymbol` method, otherwise the `evaluate` method will be used for both, literals and references so there is no real reason to test both. I'd suggest to remote them.